librz/core/cdebug.c

```diff
@@ -251,19 +238,71 @@ RZ_IPI bool rz_core_debug_reg_set(RzCore *core, const char *regname, ut64 val, c
 		rz_reg_set_value(core->dbg->reg, r, val);
 	}
 	rz_debug_reg_sync(core->dbg, RZ_REG_TYPE_ANY, true);
-	rz_core_debug_regs2flags(core, bits);
+	rz_core_debug_regs2flags(core);
 	return true;
 }
 
-RZ_IPI bool rz_core_debug_reg_list(RzCore *core, int type, int size, PJ *pj, int rad, const char *use_color) {
+static bool foreach_reg_cb(RzIntervalNode *node, void *user) {
+	RzRegItem *from_list = user;
+	RzRegItem *from_tree = node->data;
+	if (from_list == from_tree) {
+		return true;
+	}
+	// Check if from_list is covered entirely by from_tree, but is also smaller than it.
+	// We already know that
+	//   from_tree->offset <= from_list->offset < from_tree->offset + from_tree->size
+	if (from_list->offset + from_list->size > from_tree->offset + from_tree->size) {
+		// from_list expands beyond from_tree, so it's not covered
+		return true;
+	}
+	if (from_list->offset + from_list->size == from_tree->offset + from_tree->size) {
+		// they end at the same position, so it is covered entirely, but is it also smaller?
+		if (from_list->offset == from_tree->offset) {
+			// nope
+			return true;
+		}
+	}
+	// from_list ends before from_tree, so it is covered and smaller
+	return false;
+}
+
+/// Filter out all registers that are smaller than but covered entirely by some other register
+static RZ_OWN RzList *regs_filter_covered(RZ_BORROW const RzList /* <RzRegItem> */ *regs) {
+	RzList *ret = rz_list_new();
+	if (!ret) {
+		return NULL;
+	}
+	RzIntervalTree t;
+	rz_interval_tree_init(&t, NULL);
+	RzRegItem *item;
+	RzListIter *it;
+	rz_list_foreach (regs, it, item) {
+		if (item->offset < 0 || item->size <= 0) {
+			continue;
+		}
+		rz_interval_tree_insert(&t, item->offset, item->offset + item->size - 1, item);
+	}
+	rz_list_foreach (regs, it, item) {
+		if (item->offset < 0 || item->size <= 0) {
+			rz_list_push(ret, item);
+			continue;
+		}
+		if (!rz_interval_tree_all_in(&t, item->offset, true, foreach_reg_cb, item)) {
+			// foreach_reg_cb break-ed so it found a cover
+			continue;
+		}
+		rz_list_push(ret, item);
+	}
+	rz_interval_tree_fini(&t);
+	return ret;
+}
+
+RZ_IPI bool rz_core_debug_reg_list(RzCore *core, int type, int size, bool skip_covered, PJ *pj, int rad, const char *use_color) {
```

Comment:
What about killing `rad` and use `RzOutputMode`, or even better `RzCmdStateOutput`, to avoid passing `pj`, and adding table mode in the future. Moreover, some of the options like `skip_covered` and `use_color` could be combined in some kind of structure.

Reply:
Agreed, there is definitely more to improve here. Also converting `ar`/`dr` to newshell and merging their code into one.